server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -208,12 +224,15 @@ protected TransportReplicationAction(
             this::handlePrimaryRequest
         );
 
-        // we must never reject on because of thread pool capacity on replicas
+        boolean canTripCircuitBreakerOnReplica = switch (replicaActionExecution) {
+            case BypassCircuitBreaker -> false;
+            case SubjectToCircuitBreaker -> true;
+        };
         transportService.registerRequestHandler(
             transportReplicaAction,
             executor,
-            true,
-            true,
+            true, // we must never reject because of thread pool capacity on replicas
+            canTripCircuitBreakerOnReplica,
```

Comment:
No @DaveCTurner, the signature for `TransportService#registerRequestHandler` is as follows

https://github.com/elastic/elasticsearch/blob/e2129dede03e228ab45e91277e4430c12b07b688/server/src/main/java/org/elasticsearch/transport/TransportService.java#L1228-L1245

That fourth parameter is `canTripCircuitBreaker` which _currently_ defaults to true on main. I guess it was deemed more important to avoid OOM than to fail the shard?

#109414 overrides this specifically for the retention lease sync actions because we know they're very unlikely to cause an OOM. Or at least that has been my understanding.

Reply:
Ah yeah sorry you're right, those two `true` parameters have opposite senses, one means "push ahead with this request ignoring the queue length limit" and the other one means "don't push ahead with this request if the CB says not to".

That's pretty much exactly the problem I was hoping to solve by using enums here, see https://github.com/elastic/elasticsearch/pull/111017#discussion_r1682461655.

Reply:
Yep makes a lot of sense. If all remains quiet for this unplanned work iteration, I may even have a go at it (provided it doesn't turn into a can of worms)